### PR TITLE
[AVCe][lib] Add rounding offset support for AVC encoder.

### PR DESCRIPTION
**Pull request by dlin2**

### CODE REVIEW

File: _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp [hunk 1]
```diff
@@ -6334,6 +6337,34 @@ mfxStatus MfxHwH264Encode::CheckRunTimeExtBuffers(
             checkSts = sts;
     }
 
+    for (mfxU16 fieldId = 0; fieldId < NumFields; fieldId++)
+    {
+        mfxExtAVCRoundingOffset* extRoundingOffset =
+                reinterpret_cast<mfxExtAVCRoundingOffset*>(MfxHwH264Encode::GetExtBuffer(ctrl->ExtParam, ctrl->NumExtParam, MFX_EXTBUFF_AVC_ROUNDING_OFFSET, fieldId));
+        if (extRoundingOffset)
+        {
+            if (extRoundingOffset->EnableRoundingIntra == MFX_CODINGOPTION_ON)
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp]: IsOn function does the same

File: _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp [hunk 2]
```diff
@@ -6334,6 +6337,34 @@ mfxStatus MfxHwH264Encode::CheckRunTimeExtBuffers(
...
+            {
+                if (extRoundingOffset->RoundingOffsetIntra < 0 ||
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp]: As I see from API, this field is declared unsigned

File: _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp [hunk 3]
```diff
@@ -6334,6 +6337,34 @@ mfxStatus MfxHwH264Encode::CheckRunTimeExtBuffers(
...
+                        extRoundingOffset->RoundingOffsetIntra > 7)
+                {
+                    checkSts = MFX_WRN_INCOMPATIBLE_VIDEO_PARAM;
+                    extRoundingOffset->RoundingOffsetIntra = extRoundingOffset->RoundingOffsetIntra % 7;
+                }
+            }
+
+            if (extRoundingOffset->EnableRoundingInter == MFX_CODINGOPTION_ON)
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp]: same

File: _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp [hunk 4]
```diff
@@ -6334,6 +6337,34 @@ mfxStatus MfxHwH264Encode::CheckRunTimeExtBuffers(
...
+            {
+                if (extRoundingOffset->RoundingOffsetInter < 0 ||
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp]: same

File: _studio/mfx_lib/shared/include/mfx_h264_encode_vaapi.h [hunk 5]
```diff
@@ -238,6 +243,7 @@ namespace MfxHwH264Encode
         VABufferID m_rirId;                     // VAEncMiscParameterRIR
         VABufferID m_qualityParamsId;           // VAEncMiscParameterEncQuality
         VABufferID m_miscParameterSkipBufferId; // VAEncMiscParameterSkipFrame
+        VABufferID m_roundingOffsetId;          // VAEncMiscParameterCustomRoundingControl
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/include/mfx_h264_encode_vaapi.h]: Please, add initialization of this field to constructor

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp [hunk 6]
```diff
@@ -699,6 +699,69 @@ static mfxStatus SetROI(
...
+    mfxExtAVCRoundingOffset const & roundingOffset,
+    VABufferID & roundingOffsetBuf_id)
+{
+    VAStatus vaSts;
+    VAEncMiscParameterBuffer *misc_param;
+    VAEncMiscParameterCustomRoundingControl *rounding_param;
+
+    if ( roundingOffsetBuf_id != VA_INVALID_ID)
+    {
+        vaDestroyBuffer(vaDisplay, roundingOffsetBuf_id);
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: vaDestroyBuffer returns status as well, so you have to check it

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp [hunk 7]
```diff
@@ -699,6 +699,69 @@ static mfxStatus SetROI(
...
+
+    vaSts = vaMapBuffer(vaDisplay,
+                 roundingOffsetBuf_id,
+                (void **)&misc_param);
+    MFX_CHECK_WITH_ASSERT(VA_STATUS_SUCCESS == vaSts, MFX_ERR_DEVICE_FAILED);
+
+    misc_param->type = (VAEncMiscParameterType)VAEncMiscParameterTypeCustomRoundingControl;
+
+    rounding_param = (VAEncMiscParameterCustomRoundingControl *)misc_param->data;
+    if (roundingOffset.EnableRoundingIntra == MFX_CODINGOPTION_ON)
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: IsOn

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp [hunk 8]
```diff
@@ -699,6 +699,69 @@ static mfxStatus SetROI(
...
+        rounding_param->rounding_offset_setting.bits.enable_custom_rouding_intra  = 1;
+        rounding_param->rounding_offset_setting.bits.rounding_offset_intra        = roundingOffset.RoundingOffsetIntra;
+    }
+    else
+    {
+        rounding_param->rounding_offset_setting.bits.enable_custom_rouding_intra  = 0;
+        rounding_param->rounding_offset_setting.bits.rounding_offset_intra        = 0;
+    }
+
+    if (roundingOffset.EnableRoundingInter == MFX_CODINGOPTION_ON)
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: same

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp [hunk 9]
```diff
@@ -2733,6 +2800,12 @@ mfxStatus VAAPIEncoder::Execute(
         }
     }
 
+    if (m_caps.RoundingOffset && ctrlRoundingOffset)
+    {
+        MFX_CHECK_WITH_ASSERT(MFX_ERR_NONE == SetRoundingOffset(m_vaDisplay, m_vaContextEncode, *ctrlRoundingOffset, m_roundingOffsetId), MFX_ERR_DEVICE_FAILED);
+
+        configBuffers[buffersCount++] = m_roundingOffsetId;
```
Comment (mgonchar) [inline _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: are you sure that the size of configBuffer is enough to hold new buffer? I think you need to extend MAX_CONFIG_BUFFERS_COUNT

File: _studio/mfx_lib/shared/include/mfx_h264_encode_vaapi.h (hunk 5 above)
Comment (dlin2) [reply _studio/mfx_lib/shared/include/mfx_h264_encode_vaapi.h]: OK, will add it.

File: _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp (hunk 1 above)
Comment (dlin2) [reply _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp]: OK

File: _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp (hunk 2 above)
Comment (dlin2) [reply _studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp]: yes, it's unsigned, just want to show the range explicitly here. will remove it.

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp (hunk 6 above)
Comment (dlin2) [reply _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: OK

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp (hunk 7 above)
Comment (dlin2) [reply _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: OK

File: _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp (hunk 9 above)
Comment (dlin2) [reply _studio/mfx_lib/shared/src/mfx_h264_encode_vaapi.cpp]: OK

**Pull request merged**